File: core/src/dc/targetman/epf/graphics/EntityGraphDrawer.java
package dc.targetman.epf.graphics;

import java.util.ArrayList;
import java.util.List;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer.ShapeType;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector2;

import dc.targetman.ai.DefaultNode;
import dc.targetman.epf.parts.AiPart;
import dclib.epf.Entity;
import dclib.epf.graphics.EntityDrawer;
import dclib.epf.parts.TransformPart;
import dclib.geometry.RectangleUtils;
import dclib.graphics.ScreenHelper;

public final class EntityGraphDrawer implements EntityDrawer {

	private final ShapeRenderer shapeRenderer;
	private final ScreenHelper screenHelper;

	public EntityGraphDrawer(final ShapeRenderer shapeRenderer, final ScreenHelper screenHelper) {
		this.shapeRenderer = shapeRenderer;
		this.screenHelper = screenHelper;
	}

	@Override
	public final void draw(final List<Entity> entities) {
		screenHelper.setScaledProjectionMatrix(shapeRenderer);;
		shapeRenderer.setColor(Color.CYAN);
		shapeRenderer.begin(ShapeType.Line);
		for (Entity entity : entities) {
			draw(entity);
		}
		shapeRenderer.end();
	}

	private void draw(final Entity entity) {
		AiPart aiPart = entity.tryGet(AiPart.class);
		if (aiPart != null && !aiPart.getPath().isEmpty()) {
			List<Vector2> path = toVectors(aiPart.getPath());
			Rectangle bounds = entity.get(TransformPart.class).getTransform().getBounds();
			Vector2 start = RectangleUtils.base(bounds);
			Vector2 end = path.get(0);
			for (int i = 1; i < path.size(); i++) {
				shapeRenderer.line(start, end);
				start = path.get(i - 1);
				end = path.get(i);
			}
		}
	}

	private List<Vector2> toVectors(final List<DefaultNode> path) {
		List<Vector2> vectors = new ArrayList<Vector2>();
		for (DefaultNode node : path) {
			vectors.add(node.getPosition());
		}
		return vectors;
	}

}

File: core/src/dc/targetman/epf/graphics/EntityGraphDrawer.kt
package dc.targetman.epf.graphics

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.graphics.glutils.ShapeRenderer.ShapeType
import dc.targetman.epf.parts.AiPart
import dclib.epf.Entity
import dclib.epf.graphics.EntityDrawer
import dclib.epf.parts.TransformPart
import dclib.geometry.base
import dclib.graphics.ScreenHelper

class EntityGraphDrawer(private val shapeRenderer: ShapeRenderer, private val screenHelper: ScreenHelper)
    : EntityDrawer {
    override fun draw(entities: List<Entity>) {
        screenHelper.setScaledProjectionMatrix(shapeRenderer)
        shapeRenderer.color = Color.CYAN
        shapeRenderer.begin(ShapeType.Line)
        for (entity in entities) {
            draw(entity)
        }
        shapeRenderer.end()
    }

    private fun draw(entity: Entity) {
        val aiPart = entity.tryGet(AiPart::class.java)
        if (aiPart != null && !aiPart.path.isEmpty()) {
            val path = aiPart.path.map { it.position }
            val bounds = entity.get(TransformPart::class.java).transform.bounds
            var start = bounds.base
            var end = path[0]
            for (i in 1..path.size - 1) {
                shapeRenderer.line(start, end)
                start = path[i - 1]
                end = path[i]
            }
        }
    }
}
